version: "3.8"
services:

  # base images
  flask_service:
    build:
      context: ./shared/flask
    image: flask_service:python-3.10-slim-bullseye
    restart: "no"
    deploy:
      replicas: 0

  tracing:
    build:
      context: ./shared/tracing
    image: envoyproxy:tracing
    restart: "no"
    deploy:
      replicas: 0

  # jaeger
  front-envoy:
    image: envoyproxy/envoy-alpine:v1.21.6
    volumes:
      - ./front-envoy-jaeger.yaml:/etc/front-envoy.yaml
    command: "/usr/local/bin/envoy -c /etc/front-envoy.yaml --service-cluster front-proxy"
    ports:
      - "8000:8000"

  service1:
    build:
      context: ./shared/tracing
    volumes:
      - ./service1-envoy-jaeger.yaml:/etc/service-envoy.yaml
    environment:
      - SERVICE_NAME=1

  service2:
    build:
      context: ./shared/tracing
    volumes:
      - ./service2-envoy-jaeger.yaml:/etc/service-envoy.yaml
    environment:
      - SERVICE_NAME=2

  jaeger:
    image: jaegertracing/all-in-one@sha256:af062c6d0632f5bcf82a8a5572442fec2c79af7a780d689e9f1dbe17f4544fe2
    environment:
      - COLLECTOR_ZIPKIN_HOST_PORT=9411
    ports:
      - "16686:16686"